### PR TITLE
Test: Refactoring the test setup for integration, domain and system tests

### PR DESCRIPTION
**Pull request by prtandrup**
## Description

A now merged [PR](https://github.com/Energinet-DataHub/geh-charges/pull/1533) had a few unsolved comments, which are solved in this PR. 

Basically this PR addresses the following comments:  
* Make it obvious to the next developer where `B2cTenantId` and `BackendAppId` and is coming from when  ChargesFunctionAppFixture is initialized - [comment](https://github.com/Energinet-DataHub/geh-charges/pull/1533#discussion_r942151394).
* Avoid violating SOLID in `AuthorizedTestActor` - [comment](https://github.com/Energinet-DataHub/geh-charges/pull/1533#discussion_r942208524)
* Expose test actor properties in the fixture to reduce lines of code in test classes - [comment](https://github.com/Energinet-DataHub/geh-charges/pull/1533#discussion_r942202265) 

**Tasks:**

- [x] Refactor `AuthorizationConfiguration` so it has a list of (test)Clients, that each share the same properties, to avoid doing `.First()` in `ChargesFunctionAppFixture` like shown here <br>

`Environment.SetEnvironmentVariable(EnvironmentSettingNames.B2CTenantId, AuthorizationConfigurations.First().B2cTenantId);`

- [x] Rename `AuthorizedHttpRequestGenerator` to `AuthorizedTestActor` and refactor it, so it does not violate SOLID by doing both authentication and prepare http post requests. The later can be done on another existing class `HttpRequestGenerator` that is called from `AuthorizedTestActor`

- [x] In `ChargesFunctionAppFixture`, expose a property for each TestActor, i.e. our single GridAccessProvider and System Operator and let the test classes call them, something like this `Fixture.AsGridAccessProvider.PreprareHttpPostRequest()`

## References

* #1204

### CODE REVIEW

**Review by dstenroejl — COMMENTED** [review]
Nice work. :)
I added some comments/suggestions.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/IntegrationTests/WebApi/SwaggerTests.cs [hunk 1]
```diff
@@ -44,7 +45,7 @@ public class SwaggerTests :
             _client = factory.CreateClient();
             _authenticationClient = new BackendAuthenticationClient(
                 chargesWebApiFixture.AuthorizationConfiguration.BackendAppScope,
-                chargesWebApiFixture.AuthorizationConfiguration.ClientCredentialsSettings,
+                chargesWebApiFixture.AuthorizationConfiguration.TestClients.First(tc => tc.ClientName == AuthorizationConfigurationData.GridAccessProvider8100000000030).ClientCredentialsSettings,
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/IntegrationTests/WebApi/SwaggerTests.cs:48]:
```suggestion
                chargesWebApiFixture.AuthorizationConfiguration.TestClients
                    .First(tc => tc.ClientName ==AuthorizationConfigurationData.GridAccessProvider8100000000030)
                        .ClientCredentialsSettings,
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTest.Core/TestHelpers/TestClient.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using GreenEnergyHub.Charges.IntegrationTest.Core.Authorization;
+
+namespace GreenEnergyHub.Charges.IntegrationTest.Core.TestHelpers
+{
+    public class TestClient
```
[the remark this answers is not in the document]
Comment (prtandrup) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTest.Core/TestHelpers/TestClient.cs]: Went with B2CTestClient, in lack of creativity 😄 https://github.com/Energinet-DataHub/geh-charges/pull/1550/commits/0279fb86ff2e847c8c35d67748646907bb2c77b9

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTest.Core/TestHelpers/TestClient.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        /// <summary>
+        /// Retrieve B2C test client settings necessary for acquiring an access token for a given 'test client app' in the configured environment.
+        /// </summary>
+        /// <param name="clientName">Team name or shorthand.</param>
+        /// <param name="clientId">Client ID</param>
+        /// <param name="clientSecret">Client secret</param>
+        /// <returns>Settings for 'test client app'</returns>
+        private static ClientCredentialsSettings RetrieveB2CTestClientSettings(string clientName, string clientId, string clientSecret)
+        {
+            if (string.IsNullOrWhiteSpace(clientName))
```
[the remark this answers is not in the document]
Comment (prtandrup) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTest.Core/TestHelpers/TestClient.cs]: 👍 Args are now validated in constructor in https://github.com/Energinet-DataHub/geh-charges/pull/1550/commits/053848546a034818aeaee1d8ed56ca33b2a1a29d

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTest.Core/Fixtures/FunctionApp/ChargesFunctionAppFixture.cs [hunk 4]
```diff
@@ -42,8 +42,11 @@ public ChargesFunctionAppFixture()
             AzuriteManager = new AzuriteManager();
             IntegrationTestConfiguration = new IntegrationTestConfiguration();
             DatabaseManager = new ChargesDatabaseManager();
-            AuthorizationConfigurations = CreateAuthorizationConfigurations();
-            AuthorizedHttpRequestGenerators = CreateAuthorizedHttpRequestGenerators(AuthorizationConfigurations);
+            AuthorizationConfiguration = CreateAuthorizationConfiguration();
+            AuthorizedTestActors = CreateAuthorizedTestActors(AuthorizationConfiguration.B2CTestClients);
+            AsSystemOperator = SetTestActor(AuthorizationConfigurationData.SystemOperator);
+            AsSystemOperator = SetTestActor(AuthorizationConfigurationData.SystemOperator);
```
Comment (x-platformcoder) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTest.Core/Fixtures/FunctionApp/ChargesFunctionAppFixture.cs:48]: AsSystemOperator is double

**Review by HenrikSommer — APPROVED** [review]
Approved, please resolve the last few comments before merging

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTest.Core/Fixtures/FunctionApp/ChargesFunctionAppFixture.cs (hunk 4 above)
Comment (prtandrup) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTest.Core/Fixtures/FunctionApp/ChargesFunctionAppFixture.cs]: Fixed in https://github.com/Energinet-DataHub/geh-charges/pull/1550/commits/30265b3171257d599c3723ee5d2c7baa335330f9

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/IntegrationTests/WebApi/SwaggerTests.cs (hunk 1 above)
Comment (prtandrup) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/IntegrationTests/WebApi/SwaggerTests.cs]: Fixed in https://github.com/Energinet-DataHub/geh-charges/pull/1550/commits/8c034bfd42967db41e718d661edf90050b289c73

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTest.Core/TestHelpers/AuthorizedTestActor.cs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    public class AuthorizedTestActor
+    {
+        private readonly string _localTimeZoneName;
+
+        public TestClient TestClient { get; }
+
+        private AuthenticationResult? AuthenticationResult { get; set; }
+
+        public AuthorizedTestActor(TestClient testClient, string localTimeZoneName)
+        {
```
[the remark this answers is not in the document]
Comment (prtandrup) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTest.Core/TestHelpers/AuthorizedTestActor.cs:32]: Fixed in https://github.com/Energinet-DataHub/geh-charges/pull/1550/commits/868d30bd9c9190d3b549ebb403089440fc442646

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTest.Core/Authorization/AuthorizationConfiguration.cs [hunk 6]
```diff
@@ -95,26 +95,30 @@ public class AuthorizationConfiguration
...
-        /// <returns>Settings for 'team client app'</returns>
-        public static ClientCredentialsSettings RetrieveB2CTeamClientSettings(string team, string clientId, string clientSecret)
+        /// <param name="clientNames">List of team names or shorthands</param>
+        /// <returns>A list of test clients apps</returns>
+        /// <exception cref="ArgumentException">When string value is null or whitespace</exception>
+        private IEnumerable<TestClient> CreateTestClients(IEnumerable<string> clientNames)
         {
-            if (string.IsNullOrWhiteSpace(team))
-                throw new ArgumentException($"'{nameof(team)}' cannot be null or whitespace.", nameof(team));
+            ArgumentNullException.ThrowIfNull(clientNames);
```
[the remark this answers is not in the document]
Comment (prtandrup) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTest.Core/Authorization/AuthorizationConfiguration.cs]: Fixed in https://github.com/Energinet-DataHub/geh-charges/pull/1550/commits/85c8d5ee5ce27bbdf865633f1bf1b39f2070ee6c

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTest.Core/Fixtures/FunctionApp/ChargesFunctionAppFixture.cs [hunk 7]
```diff
@@ -298,31 +301,32 @@ protected override async Task OnDisposeFunctionAppDependenciesAsync()
             await DatabaseManager.DeleteDatabaseAsync();
         }
 
-        private static IEnumerable<AuthorizationConfiguration> CreateAuthorizationConfigurations()
+        private static AuthorizationConfiguration CreateAuthorizationConfiguration()
```
[the remark this answers is not in the document]
Comment (prtandrup) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTest.Core/Fixtures/FunctionApp/ChargesFunctionAppFixture.cs]: Calling inner method in https://github.com/Energinet-DataHub/geh-charges/pull/1550/commits/3323e4acb0df709d2f67db31ad6904aafbe42a34

**Pull request merged**